### PR TITLE
time: arm the remaining 0-means-unset stamps through the helpers

### PR DESCRIPTION
**Pull request by thebentern**
Follow-up to #11692, and **stacked on it** - it needs the `Time::skipZero()` / `Time::timerEndsAtMillis()` helpers and the lint rule that PR adds, so please merge that one first. The diff below this PR's own commits is #11692.

A field here uses `0` to mean "unset" - some read spells `if (f)`, `f != 0`, `f == 0 ||` or `f > 0`, or a site disarms it with `f = 0` - but it was armed from a raw clock read. Once per ~49.7-day wrap it therefore stores the exact value its own readers treat as never-set. **34 fields, 58 arm sites.**

## The rule could not have found most of these

It treated `field = <variable>` as inheriting whatever that variable did, which made the commonest shape in the tree invisible:

```cpp
unsigned long now = millis();   // top of runOnce()
...
pressStartTime = now;           // never flagged
```

Listing those names in `SENTINELS` would have bought **no protection at all**. So the scanner now tracks a local assigned from a clock and treats a write from it as the raw arm it is. One hop, one function, name-based; it forgets a local reassigned from anything else, and taint is dropped at each function boundary. Twelve fixtures pin it, including the negative cases: no leak across functions, `now` does not match `nowMs`, and neither `==` nor `+=` records anything.

That pass immediately found a site #11692 itself missed - `setTransmitDelay()` recomputing `p->tx_after` from a tainted `now`, two lines under the `if (p->tx_after)` read that takes 0 as "no delay wanted". That fix went into #11692 where the review raised it independently.

## Three worth naming, because the consequence is not cosmetic

| Field | What happens on the wrap tick |
| --- | --- |
| `UpDownInterruptBase` press/up/down`StartTime` | `xDetected` is only cleared **inside** the block guarded by `xDetected && xStartTime > 0`, so a stored 0 makes both the entry and the exit condition unreachable. That button is dead for the rest of the boot, not for one tick. |
| `PhoneAPI::lastContactMsec` | `ServerAPI` reads `lastContactMsec > 0` before the TCP idle close, and the field stays 0 until the next inbound packet - so a client that never speaks again leaks the socket for the life of the connection. |
| `EInkDisplay::lastDrawMsec` | `if (lastDrawMsec)` gates every plain `display()` call on a keyframe having been shown, so a stored 0 stops the screen updating until something calls `Screen::forceDisplay()` again. |

## TransmitHistory needed more than its arm sites

`getLastSentToMeshMillis()` returns `0` to mean "module has never sent". Besides the two stores, **both** reconstruction helpers end in `millis() - msAgo` and can produce a 0 of their own, so guarding only the stores would have left two paths broken. All three computed returns are guarded; the deliberate `return 0;` sentinels are untouched.

## Judged and deliberately not changed

* **`connect_time_ms`** (nRF54L15) is armed from `k_uptime_get_32()`, not `millis()`. It is guarded with `skipZero()` but **keeps its own clock** - swapping in `Time::getMillis()` would have it compared against a `k_uptime` now at the watchdog read. The rule now recognises that clock too, so listing the field is not an empty gesture.
* **`RotaryEncoderInterruptBase::pressStartTime`** shares a name with the UpDown field and has a *different contract*: no read here tests the stamp against 0, `pressDetected` is the only armed flag. Opted out at the write. Its `lastPressLongEventTime` sibling **is** a `== 0` latch and is fixed.
* **`PositionModule:38`** copies a value the enclosing `if (restored != 0)` has already proven non-zero. Opted out.
* **`pmMeasureStarted`**, **`adminKeyFallbackRefillMs`**, the two autosave stamps and **`scrollStartDelay`** are lazy initialisations whose wrap behaviour costs at most one interval and drops nothing. Left alone, and not listed.

## How the list was arrived at

Each candidate was mapped (every read and write, storage class, companion flag, stamp-versus-deadline shape) and then classified from three independent angles, because a wrong "exempt" writes a permanent suppression over a real bug and a wrong "fix" churns working code. 33 came back unanimous; the two split cases were read by hand - `fileage` fixed, `pmMeasureStarted` left alone.

## Testing

Full native suite **1421/1421, 78 suites, 0 compile errors**. Lint self-test 53 fixtures, green under bash 3.2 as well. The rule is silent across `src/`. `trunk fmt` clean.

## 🤝 Attestations

- [ ] I have tested that my proposed changes behave as described.
- [ ] I have tested that my proposed changes do not cause any obvious regressions on the following devices:
  - [ ] Heltec (Lora32) V3
  - [ ] LilyGo T-Deck
  - [ ] LilyGo T-Beam
  - [ ] RAK WisBlock 4631
  - [ ] Seeed Studio T-1000E tracker card
  - [ ] Other (please specify below)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved timer and timeout reliability across uptime-counter wraparound.
  - Updated display refresh, input handling, GPS, networking, telemetry, MQTT, and power-management timing to avoid treating a valid timestamp as “unset.”
  - Improved deadline handling for temporary UI messages, reboot delays, throttling, and route freshness checks.
  - Prevented timing-related issues at the counter’s zero-millisecond boundary.

- **Tests**
  - Added coverage for zero-safe timestamps, deadline calculations, counter wraparound, and timing edge cases.

### CODE REVIEW

File: src/UptimeClock.h [hunk 1]
```diff
@@ -44,6 +44,30 @@ void setMonotonicPublishHookForTests(MonotonicPublishHook hook);
 /// which is wrap-correct with no carry state at all.
 uint32_t getMillis();
 
+/// Step a millis value past 0. Stored stamps and deadlines conventionally use 0 for "unset", so the
+/// one tick per ~49.7-day wrap that lands on 0 would read as never-set; 1 is a 1 ms error instead.
+constexpr uint32_t skipZero(uint32_t ms)
+{
+    return ms ? ms : 1;
```
[the remark this answers is not in the document]
Comment (thebentern) [reply src/UptimeClock.h:51]: You are right, and this was the substantive one - fixed in 7320e006e.

Dodging at the store while a reader measures against a raw clock splits the two sides apart for one tick per wrap: the stamp becomes 1 while `now` is still 0, so `now - stamp` is UINT32_MAX and a brand-new stamp reads as ~49.7 days old. I confirmed two concrete consequences before changing anything - `UpDownInterruptBase` computed `now - pressStartTime` and emitted a long press for a *fresh* press, and `TraceRouteModule` read `now - lastTraceRouteTime < cooldownMs` as false and bypassed its cooldown.

Rather than a separate elapsed-time helper, the dodge moved to the read: `Time::stampMillis()` is `getMillis()` with the one 0 tick called 1. A site that both stores a stamp and measures against stamps reads the clock once through it and stores that value directly, so both sides are on one value by construction. Nine files.

Where the clock arrives as a **parameter** the store keeps its own `skipZero()` too, because the function cannot assume its caller dodged anything - and the two compose without shifting twice. Removing that was a real regression on my side: `test_nexthop_routing` caught it, because `test_health_learn_never_stores_zero_sentinel` calls `noteRouteLearned(DEST, 0xAB, 0)` directly and asserts the store normalises it (0 is the empty-slot marker `getOrAllocRouteHealth()` evicts on). `trySmartBroadcast()` and `directResponseAllowed()` have the same shape and keep their guard for the same reason.

Five tests added in `test_uptime_clock`, including one that asserts the old store-only shape really does yield UINT32_MAX, and one pinning the 1 ms skew at 399 rather than 400 so nobody "corrects" it back into a raw read on one side.

File: src/input/UpDownInterruptBase.cpp [hunk 2]
```diff
@@ -60,13 +61,13 @@ int32_t UpDownInterruptBase::runOnce()
...
-        pressStartTime = now;
+        pressStartTime = Time::skipZero(now);
     } else if (this->action == UPDOWN_ACTION_UP && upButtonPressed && !upDetected) {
         upDetected = true;
-        upStartTime = now;
+        upStartTime = Time::skipZero(now);
     } else if (this->action == UPDOWN_ACTION_DOWN && downButtonPressed && !downDetected) {
         downDetected = true;
-        downStartTime = now;
+        downStartTime = Time::skipZero(now);
```
[the remark this answers is not in the document]
Comment (thebentern) [reply src/input/UpDownInterruptBase.cpp]: Fixed in 7320e006e. `now` is read through `Time::stampMillis()` at the top of `runOnce()` and all six stamps store it directly, so the duration arithmetic and the stored value are the same number - `now - pressStartTime` is 0 on the wrap tick instead of UINT32_MAX, and the spurious long press is gone.

One thing worth recording that is NOT in scope here: `now` and `lastPressKeyTime`/`lastUpKeyTime`/`lastDownKeyTime` are all `unsigned long`, which is 64-bit on the portduino host, so the three debounce comparisons are not 32-bit modular and fail open once per wrap. The six sentinel durations are immune because they are captured into `uint32_t` first. That width issue predates this PR - `git show` confirms the same declarations before the change - so I have left it alone rather than widen the diff.

File: src/mesh/NextHopRouter.cpp [hunk 3]
```diff
@@ -617,7 +617,7 @@ void NextHopRouter::noteRouteLearned(NodeNum dest, uint8_t nextHop, uint32_t now
         h->lastNextHop = nextHop;
         h->consecutiveFailures = 0;
     }
-    h->learnedAtMsec = now ? now : 1;
+    h->learnedAtMsec = Time::skipZero(now);
```
[the remark this answers is not in the document]
Comment (thebentern) [reply src/mesh/NextHopRouter.cpp:621]: Fixed in 7320e006e. `doRetransmissions()` reads through `Time::stampMillis()`, and the three route-health call sites that previously passed a raw `Time::getMillis()` - the two `isRouteStale()` checks including the early-flood one, and `noteRouteLearned()` in `sniffReceived()` - now pass a normalised value, as does `noteRouteSuccess()` from `ReliableRouter`.

The stores inside `noteRouteLearned()` and `noteRouteSuccess()` also keep their own `skipZero()`, because `now` is a parameter there and those two are called with a literal 0 by `test_nexthop_routing`. I had briefly removed those guards on the theory that fixing the callers was enough; the tests correctly rejected that.

`PacketHistory::insert()` I have deliberately left - see the reply on the `TransmitHistory` thread for why the Throttle-internal read is a separate, tree-wide question.

File: bin/lint-unset-sentinel-millis.sh [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+		while (match(rest, "(" sentinels ")[ \t]*=")) {
+			p = RSTART; len = RLENGTH
+			off += p
+			# A plain assignment only: ==, !=, <=, >=, +=, -= and friends are reads or updates,
+			# not an arming write, and `=` must not be the head of `==`.
+			eq = off + len - 1
+			prev = (eq - 1 >= 1) ? substr(code, eq - 1, 1) : " "
+			nxt  = (eq + 1 <= length(code)) ? substr(code, eq + 1, 1) : " "
+			if (prev !~ /[-+*\/%&|^!<>=]/ && nxt != "=" && !ident_before(code, off) &&
+			    !is_declaration(code, off)) {
```
[the remark this answers is not in the document]
Comment (thebentern) [reply bin/lint-unset-sentinel-millis.sh:374]: Real, and not hypothetical - `src/modules/SerialModule.h:22` has `unsigned long lastNmeaTime = millis();` as a class-member initializer today. (That field is not a sentinel - no read tests it against 0 - but the shape is there.) Fixed in 7320e006e: the rule now tracks class-body scope, so `is_declaration()` only excuses a typed declaration inside a *function* body, and a member initializer reading the clock is reported.

Getting that right took two attempts, and the first one was worse than the bug. It introduced three false positives, all shapes that are not class bodies at all:

```cpp
template <class T> void f(T x) { uint32_t lastSort = millis(); }
class Foo { void tick() { uint32_t lastSort = millis(); } };
void g(struct Bar *b) { uint32_t lastSort = millis(); }
```

Two causes: `pending_class` matched `class`/`struct` anywhere on the line, so a template parameter list and an elaborated type in a parameter list both marked the following function body as class scope - it is anchored to the start of the line now; and `update_scope()` runs at the end of a line, so a body opened earlier on the *same* line had not been counted when the statement was judged - `is_declaration()` now also counts unmatched braces earlier in the statement. Since the rule is blocking and `template <class T>` is ordinary C++, those would have reddened files nobody touched.

Also fixed alongside: `note_taint()` never got the per-write `;` cut the judging path was given earlier in review, so on a line with two statements it learned taint from the neighbour.

Fixtures for all of it - a class-member initializer, a local in a one-line inline method, a local in a multi-line method, the template and struct-parameter shapes, and the taint-from-neighbour case. 53 -> 65 fixtures, green under bash 3.2.

File: src/mesh/TransmitHistory.cpp [hunk 5]
```diff
@@ -81,7 +82,7 @@ void TransmitHistory::loadFromDisk()
 
 void TransmitHistory::setLastSentToMesh(uint16_t key)
 {
-    lastMillis[key] = millis();
+    lastMillis[key] = Time::skipZero(Time::getMillis());
```
[the remark this answers is not in the document]
Comment (thebentern) [reply src/mesh/TransmitHistory.cpp:85]: I do not think the premise holds here, so I have left these as they are - happy to be shown wrong.

The claim is that map membership carries the unset state, so a raw 0 is safe. It does not reach the callers that way. `getLastSentToMeshMillis()` flattens "absent" and "stored 0" into the same return value:

```cpp
auto mit = lastMillis.find(key);
if (mit != lastMillis.end())
    return mit->second;          // a stored 0 returns 0
...
return 0;                        // "No stored time - module has never sent" (the comment at that line)
```

and the callers branch on exactly that: `NodeInfoModule.cpp:163`/`:167` (`lastNodeInfo && ...`), `MeshBeaconModule.cpp:624` (`lastSent == 0 || ...`), plus the telemetry modules. So a stored 0 is read as "never sent" and the throttle is bypassed until the next send - which is a longer-lived wrong answer than the one-tick window removing the dodge would trade it for.

Worth noting the accessor has **two more** paths to a spurious 0 that the store-side guard alone would not have covered: `getLastSentBootRelativeMillis()` and `getLastSentAbsoluteMillis()` both end in `millis() - msAgo`. This PR guards all three computed returns and leaves the deliberate `return 0;` sentinels untouched.

On `lastDiskSave`: its read is `lastDiskSave == 0 || !Throttle::isWithinTimespanMs(...)`, so 0 already means "save now" - with or without the dodge that tick saves, and the outcome is identical.

The real point underneath this, which I agree with and have **not** fixed: a store that dodges zero while its reader measures through a `Throttle::` helper still splits for one tick, because those helpers read the clock internally and raw. That is about eight sites tree-wide, not just these. Closing it means making `Throttle` read through the dodge - proposed on #11692 and declined there pending a caller audit - so fixing one site here would only make the tree inconsistent. It is recorded in the commit message as a known residual.

File: bin/lint-unset-sentinel-millis.sh [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+	# Does this statement read a clock directly? Matches millis(), Time::getMillis() and any wrapper
+	# whose name ends in millis, which is how almost every clock read in this tree spells itself, plus
+	# Zephyr k_uptime_get_32() - the nRF54L15 BLE code has no millis() at all and wraps at 32 bits just
+	# the same, so a sentinel armed from it needs the same guard.
+	function reads_clock(s) { return (s ~ /[Mm]illis[ \t]*\(/ || s ~ /k_uptime_get_32[ \t]*\(/) }
+
+	# Already routed through a helper that dodges 0, so it is the fix rather than the defect. Note
+	# stampMillis() belongs here even though its name ends in millis: a local read through it holds a
+	# value that is already non-zero, so a write from that local is safe and must not be flagged.
+	function is_safe_arm(s) { return (s ~ /skipZero/ || s ~ /timerEndsAtMillis/ || s ~ /stampMillis/) }
```
[the remark this answers is not in the document]
Comment (thebentern) [reply bin/lint-unset-sentinel-millis.sh:220]: Right, and this one was embarrassing - fixed in b7efa3c76.

`stampMillis()` guarantees only its own result, so arithmetic on top of it can carry a non-zero stamp straight back onto the sentinel. Confirmed concretely: `0xFFFFEC78 + 5000` is exactly `0`. That sum is precisely what `timerEndsAtMillis()` exists to dodge, and the rule was waving it through because a helper name appeared somewhere in the expression.

Worse, I had written a fixture asserting `rebootAtMsec = now + 5000` was quiet, so the self-test was pinning the hole open. That fixture is gone.

A local holding a dodged value is now tracked separately from a tainted one. It may be stored or copied straight through, but `+`/`-` applied at the **outermost** level is reported, with the message pointing at `timerEndsAtMillis()`. The check is depth-aware so the two legitimate shapes stay quiet: the operator inside `Time::skipZero(Time::getMillis() - msAgo)` is at depth 1, and the `(d == 0) ? 0 : Time::timerEndsAtMillis(d)` arming form has no top-level operator at all.

Four fixtures replace the bad one: store-through, copy one more hop, arithmetic on a dodged local, arithmetic on a direct helper call.

File: bin/lint-unset-sentinel-millis.sh [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+				if (depth > 0) depth--
+			}
+		}
+	}
+
+	# Is the statement being judged sitting directly in a class body? `opened` is the net brace
+	# count seen earlier in this same statement, which update_scope() has not applied yet: a body
+	# opened on this very line (a one-line inline method) puts the statement inside a function, not
+	# in the class body.
+	function at_class_scope(opened) { return ((depth in class_body) && opened <= 0) }
```
[the remark this answers is not in the document]
Comment (thebentern) [reply bin/lint-unset-sentinel-millis.sh]: Confirmed and fixed in b7efa3c76 - `class Foo { uint32_t lastSort = millis(); };` was silent, and it is reported now.

Both halves were wrong, as you describe. The header check rejected the line because it ends in a semicolon, which a one-liner body always does, so it now arms on the presence of a brace instead. And even once armed, `braces_before()` counted the class brace as if it opened a function body: when the body opened on the statement being judged, one unmatched brace is class scope while two or more is a method body inside it, so `class Foo { void tick() { uint32_t lastSort = millis(); } };` still correctly excuses its local.

Getting that wrong first broke every multi-line class - setting the per-statement flag without also arming `pending_class` meant `update_scope()` never registered the body, so members stopped being reported entirely. The three existing class fixtures caught it immediately, which is the argument for having added them.

Your exact one-line fixture is in the suite now, alongside the nested-method variant and a forward-declaration case. 72 fixtures total.

**Review by NomDeTom — APPROVED** [review]
